### PR TITLE
tkt-78015: Switch from SSSD to NSLCD

### PR DESCRIPTION
**Pull request by anodos325**
- Begin using NSLCD for authenticated LDAP. More work is needed to bring in / validate kerberos support post-sssd, and test edge cases.

### CODE REVIEW

**Comment by yocalebo** [comment]
Getting rid of SSSD is amazing!

File: src/freenas/etc/directoryservice/LDAP/ctl [hunk 1]
```diff
@@ -7,6 +7,7 @@ status_file="/var/run/directoryservice.ldap"
 service=/usr/sbin/service
 python=/usr/local/bin/python
 notifier=/usr/local/bin/midclt
+smbpasswd=/usr/local/bin/smbpasswd
```
Comment (william-gr) [inline src/freenas/etc/directoryservice/LDAP/ctl]: where is this used?

File: src/middlewared/middlewared/plugins/smb.py [hunk 2]
```diff
@@ -98,6 +98,25 @@ def check_codec(encoding):
...
         return encodings
 
+    @private
+    async def store_ldap_admin_password(self):
+        """
+        This is required if the LDAP directory service is enabled. The ldap admin dn and
+        password are stored in private/secrets.tdb file.
+        """
+        ldap = await self.middleware.call('datastore.config', 'directoryservice.ldap')
+        password_stored = await self.middleware.call('cache.has_key', 'SMB_LDAP_BINDPW_SET')
```
Comment (william-gr) [inline src/middlewared/middlewared/plugins/smb.py]: Shouldnt we pop SMB_LDAP_BINDPW_SET if ldap_bindpw changes?

File: src/freenas/etc/directoryservice/LDAP/ctl (hunk 1 above)
Comment (anodos325) [reply src/freenas/etc/directoryservice/LDAP/ctl]: I was originally going to set the ldap secret inside the ctl script, but then had a brief spurt of sanity. Will delete that line.

File: src/middlewared/middlewared/plugins/smb.py (hunk 2 above)
Comment (anodos325) [reply src/middlewared/middlewared/plugins/smb.py]: After I wrote this I thought of a few other edge cases where we might end up with secrets.tdb getting out of sync. For now (until we have a proper LDAP plugin) it's probably best to just reset the password every time we run the ctl script. It adds at most a second to the service start.

**Comment by william-gr** [comment]
Thanks, looking good. Could you please rebase this branch to resolve conflicts before approval? (Actual rebase instead of merge master would be appreciated to pet my OCD of cleaner history : )

**Comment by CrossEyeORG** [comment]
Maybe I missing something but in my mind, isn't moving from SSSD to NSLCD a step backwards / a legacy approach?

What is the reasoning for moving away from SSSD, outdated ports?

From my observations, a consistent theme in the FreeNAS forums is FreeNAS has a bad LDAP/AD integration and I do not want that void to grow with this change. When I used to use NSLCD back in the day, it was painful with lots of custom scripts to get the necessary functionality.

**Comment by yocalebo** [comment]
> Maybe I missing something but in my mind, isn't moving from SSSD to NSLCD a step backwards / a legacy approach?
> 
> What is the reasoning for moving away from SSSD, outdated ports?
> 
> From my observations, a consistent theme in the FreeNAS forums is FreeNAS has a bad LDAP/AD integration and I do not want that void to grow with this change. When I used to use NSLCD back in the day, it was painful with lots of custom scripts to get the necessary functionality.

No, SSSD is a pure linux based application that is severely out-dated in the freeBSD ports tree. It has become a non-trivial task to update the SSSD port in freeBSD because the SSSD team has diverged from making that code portable (redhat, now IBM).

Also, SSSD is fundamentally broken when you deal with LDAP customers that have many 10's of thousands of user accounts. SSSD spins in mmap() calls forever. Would an update of the SSSD port fix this particular issue? Maybe, however, nslcd is perfectly capable of competing with SSSD for the use case that freeNAS and trueNAS are using it for.

**Comment by yocalebo** [comment]
I will not contest you on the "AD/LDAP" integration that has been previously seen. However, there is a new developer in town that is making incredible progress in this area to clean it up and make it far less painful. Remember, trueNAS is the enterprise product and a lot of the pain points that are felt are also felt at the enterprise level. 11.2 and further versions are greatly closing that void. Switching between SSSD and nslcd is a small step in that direction :+1:

**Comment by CrossEyeORG** [comment]
That is definitely a fair point, with Red Hat/IBM... which is understandably frustrating!

I was going to say that updating the port might be easier process than completely ditching it for something else but after your second response... I am optimistically excited to see this new direction!

Thanks for the info yocalebo!
